File: 0059-spiral-matrix-ii/0059-spiral-matrix-ii.cpp
class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        int left = 0, right = n-1, top = 0, bottom = n-1;
        int val = 1;
        vector<vector<int>> ans(n, vector<int>(n));
        while(left <= right && top <= bottom){
            // move to right
            for(int i = left; i <= right; i++){
                ans[top][i] = val++;
            }
            top++;

            // move to bottom
            for(int i = top; i <= bottom; i++){
                ans[i][right] = val++;
            }
            right--;

            // move to left
            if (top<=bottom){
                for(int i = right; i >= left; i--){
                    ans[bottom][i] = val++;
                }
                bottom--;
            }

            // move to top
            if (left <= right){
                for(int i = bottom; i >= top; i--){
                    ans[i][left] = val++;
                }
                left++;
            }
        }
        return ans;
    }
};

File: 0059-spiral-matrix-ii/0059-spiral-matrix-ii.py
class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        res = [[0]*n for _ in range(n)]
        l,r,t,b=0,n,0,n
        count=0

        while l < r and t < b:
            for i in range(l,r):
                count += 1
                res[t][i] = count
            t+=1

            for i in range(t,b):
                count += 1
                res[i][r-1]=count
            r-=1

            if l<r:
                for i in range(r-1,l-1,-1):
                    count += 1
                    res[b-1][i]=count
                b-=1
            
            if t<b:
                for i in range(b-1,t-1,-1):
                    count += 1
                    res[i][l]=count
                l+=1
            
        return res